File: scripts/split-bullet-overflow.py
#!/usr/bin/env python3
"""
Split slides with 8+ bullet points into 2 slides to prevent overflow.
"""
import json
import sys
from pathlib import Path

MAX_BULLETS = 7

def split_bullets(slides):
    new_slides = []
    split_count = 0
    for slide in slides:
        content = slide.get("content", [])
        # Only split plain bullet items (not table rows or image directives)
        bullets = [c for c in content if not c.startswith("|") and not c.startswith("![")]
        non_bullets = [c for c in content if c.startswith("|") or c.startswith("![")]
        
        if len(bullets) > MAX_BULLETS:
            split_count += 1
            mid = len(bullets) // 2
            # Slide 1: first half
            slide1 = dict(slide)
            slide1["content"] = bullets[:mid] + non_bullets
            slide1["title"] = slide["title"] + "（1/2）"
            new_slides.append(slide1)
            # Slide 2: second half
            slide2 = dict(slide)
            slide2["content"] = bullets[mid:] + non_bullets
            slide2["title"] = slide["title"] + "（2/2）"
            new_slides.append(slide2)
        else:
            new_slides.append(slide)
    return new_slides, split_count

def process_file(path):
    path = Path(path)
    with open(path, encoding="utf-8") as f:
        data = json.load(f)
    original = len(data["slides"])
    data["slides"], split_count = split_bullets(data["slides"])
    if split_count > 0:
        with open(path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        print(f"✅ {path.parent.name}: split {split_count} slides ({original} → {len(data['slides'])})")
    return split_count

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "--all":
        total = 0
        for p in sorted(Path("docs").glob("**/slides-data.json")):
            total += process_file(p)
        print(f"\n📊 Total: {total} slides split")
    elif len(sys.argv) > 1:
        process_file(sys.argv[1])

File: scripts/split-bullet-overflow.ts
#!/usr/bin/env bun
/**
 * Split slides with more than MAX_BULLETS plain bullets into two slides (1/2) and (2/2).
 * Replaces scripts/split-bullet-overflow.py
 *
 * Usage:
 *   bun run scripts/split-bullet-overflow.ts --all
 *   bun run scripts/split-bullet-overflow.ts docs/my-deck/slides-data.json
 */

import { Glob } from "bun";
import { MAX_BULLET_CHARS } from "./lib/constants.js";

interface Slide {
	title?: string;
	subtitle?: string;
	content?: string[];
	code?: string;
	codeLanguage?: string;
	speakerNotes?: string;
	layout?: string;
}

interface SlideData {
	slides: Slide[];
}

function splitBullets(slides: Slide[]): {
	slides: Slide[];
	splitCount: number;
} {
	const newSlides: Slide[] = [];
	let splitCount = 0;

	for (const slide of slides) {
		const content = slide.content ?? [];
		// Only count plain bullets (not table rows or image directives)
		const bullets = content.filter(
			(c) => !c.startsWith("|") && !c.startsWith("!["),
		);
		const nonBullets = content.filter(
			(c) => c.startsWith("|") || c.startsWith("!["),
		);

		if (bullets.length > MAX_BULLET_CHARS) {
			splitCount++;
			const mid = Math.floor(bullets.length / 2);

			// Slide 1: first half — preserve all fields
			const slide1: Slide = {
				...slide,
				title: `${slide.title ?? ""}（1/2）`,
				content: [...bullets.slice(0, mid), ...nonBullets],
			};
			// subtitle/speakerNotes only on first slide
			newSlides.push(slide1);

			// Slide 2: second half — no subtitle/speakerNotes (continuation)
			const { subtitle: _sub, speakerNotes: _notes, ...rest } = slide;
			const slide2: Slide = {
				...rest,
				title: `${slide.title ?? ""}（2/2）`,
				content: [...bullets.slice(mid), ...nonBullets],
			};
			newSlides.push(slide2);
		} else {
			newSlides.push(slide);
		}
	}

	return { slides: newSlides, splitCount };
}

async function processFile(filePath: string): Promise<number> {
	const file = Bun.file(filePath);
	if (!(await file.exists())) {
		console.error(`❌ File not found: ${filePath}`);
		return 0;
	}

	try {
		const data = (await file.json()) as SlideData;
		const original = data.slides.length;
		const { slides, splitCount } = splitBullets(data.slides);

		if (splitCount > 0) {
			data.slides = slides;
			await Bun.write(filePath, JSON.stringify(data, null, 2));
			const dirName = filePath.split("/").slice(-2, -1)[0];
			console.log(
				`✅ ${dirName}: split ${splitCount} slides (${original} → ${slides.length})`,
			);

			// Update slides.count in config to reflect new slide count
			const configPath = filePath.replace(
				"slides-data.json",
				"slides.config.yaml",
			);
			if (await Bun.file(configPath).exists()) {
				try {
					const { parse: parseYaml, stringify: stringifyYaml } = await import(
						"yaml"
					);
					const configText = await Bun.file(configPath).text();
					const config = parseYaml(configText) as Record<string, unknown>;
					if (config.slides && typeof config.slides === "object") {
						(config.slides as Record<string, unknown>).count = slides.length;
					} else {
						config.slides = { count: slides.length };
					}
					await Bun.write(configPath, stringifyYaml(config));
				} catch {
					// Config update is best-effort — don't fail the split
				}
			}
		}

		return splitCount;
	} catch (err) {
		console.error(
			`❌ Error processing ${filePath}: ${err instanceof Error ? err.message : String(err)}`,
		);
		return 0;
	}
}

async function main() {
	const args = process.argv.slice(2);

	if (args[0] === "--all") {
		const glob = new Glob("docs/**/slides-data.json");
		const files = Array.from(glob.scanSync()).sort();
		let total = 0;
		for (const f of files) {
			total += await processFile(f);
		}
		console.log(`\n📊 Total: ${total} slides split`);
	} else if (args[0]) {
		await processFile(args[0]);
	} else {
		console.error("Usage:");
		console.error("  bun run scripts/split-bullet-overflow.ts --all");
		console.error(
			"  bun run scripts/split-bullet-overflow.ts docs/my-deck/slides-data.json",
		);
		process.exit(1);
	}
}

main();
